add order and align-self exercises

rarmatei opened the pull request.

NateWr commented on week-2/13-order/ignore_this_folder/custom.css:
```diff
@@ -0,0 +1,3 @@
+.solution-example {
```
I've been putting custom CSS files in a separate CSS directory outside of the exercise directory. https://github.com/NateWr/html-css-git-exercises/tree/master/css

NateWr replied:
Same with the images below. Let's keep them out of the exercises directory: https://github.com/NateWr/html-css-git-exercises/tree/master/images

NateWr commented on week-2/13-order/index.html:
```diff
@@ -0,0 +1,56 @@
+  <title>Flexbox</title>
+  <link rel="stylesheet" href="/css/normalize.css">
+  <link rel="stylesheet" href="/css/week2.css">
+	<link rel="stylesheet" href="/css/week2-flexbox.css">
```
Whitespace.

rarmatei replied:
I'll do this as part of the "prettier" PR

NateWr commented on week-2/13-order/index.html:
```diff
@@ -0,0 +1,56 @@
+      <h1 class="site-header__title">First</h1>
+    </div>
+    <ul class="countries countries--first">
+			<li class="country" id="brazil">Brazil</li>
```
I've been using 2-space tab indentation. Do you know what other modules are using? We should probably standardize everything on beautify or prettier.

rarmatei replied:
Yes, I'll add a PR after this with "prettier" as a recommended VS extensions (so it recommends installing it when somebody opens the project) and then reformat all html and css files with prettier

NateWr commented on week-2/13-order/index.html:
```diff
@@ -0,0 +1,56 @@
+			<li class="country" id="laos">Laos</li>
+			<li class="country" id="uganda">Uganda</li>
+    </ul>
+    <p>⬇️⬇️ should look like this ⬇️⬇️</p>
```
I like this idea but I think it maybe makes the exercises look a little confusing. Instead of putting the result right next to the exercise in each case, maybe we can put the result below the exercises in a separate panel?

![selection_062](https://user-images.githubusercontent.com/2306629/50685737-4b485c80-1012-11e9-8838-b4b64a8c63af.png)

(This image also shows the Ubuntu font, which most students will see since they'll be on Linux.)

rarmatei replied:
Thanks a lot for building the preview image! 

I can see how the current setup can make it confusing. I don't know if it should be at bottom though - ideally, I was looking to avoid making the student scroll back and fourth between the solution and their implementation while working on each part.

I'll try having the solution in a separate panel, but side by side rather than the bottom. We can then discuss how it looks.

NateWr commented on week-2/13-order/readme.md:
```diff
@@ -0,0 +1,13 @@
+# Item order
+
+The `flex-order` property is used to move a *specific* item to a different position in the list.
```
> ...used to move *one* item...

NateWr replied:
Is it `flex-order` or `order`?

NateWr commented on week-2/13-order/readme.md:
```diff
@@ -0,0 +1,13 @@
+
+The `flex-order` property is used to move a *specific* item to a different position in the list.
+
+> **Remember:** if you forgot how to use the `flex-order` property, you can always:
```
I've been putting hints after the description of the exercise. I feel like if a student opens an exercise they'll want to know what they have to do first? Either way, let's choose an order and be consistent.

rarmatei replied:
ok, I'll move it after 👍

NateWr commented on week-2/13-order/readme.md:
```diff
@@ -0,0 +1,13 @@
+> 2. Click on the link from "CSS-tricks"
+> ![Screenshot of how to google flexbox](/images/13-googling-flexbox.png)
+
+Exercise:
```
I've been excessively descriptive when saying what they need to do. Maybe that's not a good idea. But I'd use something like this:

> In this exercise, you'll use `order` to move the items in the list. Follow the steps below to complete this exercise.

rarmatei replied:
I agree this is much more descriptive 👍

NateWr commented on week-2/13-order/readme.md:
```diff
@@ -0,0 +1,13 @@
+> 2. Click on the link from "CSS-tricks"
+> ![Screenshot of how to google flexbox](/images/13-googling-flexbox.png)
+
+Exercise:
```
I've been excessively descriptive when saying what they need to do. Maybe that's not a good idea. But I'd use something like this:

> In this exercise, you'll use `order` to move the items in the list. Follow the steps below to complete this exercise.

NateWr commented on week-2/13-order/readme.md:
```diff
@@ -0,0 +1,13 @@
+
+Exercise:
+
+1. Open up the HTML page for this exercise in Chrome
```
By this point, the students should be comfortable running `npm run X`, which will open the page automatically. It might be confusing if we tell them to open the HTML page in Chrome -- they might think we mean actually opening the file directly.

Suggested alternative:

> 1. Look at the First list of countries. Then look at the First Result displayed at the bottom. Find the differences between the two lists.
> 2. Use the `order` property to make the First list of countries look like the First Result list.
> 3. Do the same thing for the Second and Third list.

NateWr commented on week-2/14-align-self/flexbox.css:
```diff
@@ -0,0 +1,13 @@
+}
+/* Add your own CSS code below */
+
+.countries--first {
```
Is this (and the code below it) the answer? (I always accidentally commit the answer. :laughing:)

rarmatei replied:
.....maybe 😆

NateWr commented on week-2/13-order/readme.md:
```diff
@@ -0,0 +1,13 @@
+Exercise:
+
+1. Open up the HTML page for this exercise in Chrome
+2. Make sure each list of countries looks the image below it
```
Remove the `.DS_Store` file below from the commit.

NateWr commented on week-2/14-align-self/flexbox.css:
```diff
@@ -0,0 +1,13 @@
+
+.countries--first #uganda {
+	align-self: center;
+} 
```
Another `.DS_Store` file (below). Is there a `.gitignore` rule we should add?

rarmatei replied:
yes, I'll add one 👍

NateWr commented on week-2/14-align-self/ignore_this_folder/custom.css:
```diff
@@ -0,0 +1,7 @@
+.countries--tall {
```
Same thing here with moving the CSS and images into the general directoru, out of the exercise directory.

NateWr commented on week-2/14-align-self/index.html:
```diff
@@ -0,0 +1,67 @@
+      <h1 class="site-header__title">First</h1>
+    </div>
+    <ul class="countries countries--tall countries--first">
+			<li class="country" id="brazil">Brazil</li>
```
Let's not use the `id` attribute here. For one, there are multiple instances of the same `id` on the page, which I don't think is valid. But also, let's reinforce the BEM naming structure we've been using, and use class names like `country--brazil`.

rarmatei replied:
yes, very good spot, wasn't intentional to use multiple ids with the same name

NateWr commented on week-2/14-align-self/index.html:
```diff
@@ -0,0 +1,67 @@
+			<li class="country" id="laos">Laos</li>
+			<li class="country" id="uganda">Uganda</li>
+    </ul>
+    <p>⬇️⬇️ should look like this ⬇️⬇️</p>
```
An image for the separate results if we go that way.

![align-result](https://user-images.githubusercontent.com/2306629/50686215-44bae480-1014-11e9-8392-5151d89e1734.png)

NateWr commented on week-2/14-align-self/readme.md:
```diff
@@ -0,0 +1,13 @@
+# Align self
+
+With `align-items` you move all the items at once (on the **non-main axis**).
```
Have we described the axis before this exercise? Maybe we need an exercise specifically on the axis that occurs before this?

rarmatei replied:
I was assuming we will describe it in class - I can add an exercise in that case. Something to indicate how `justify-content` and `align-items` are linked to `flex-direction`

NateWr replied:
:+1:

NateWr commented on week-2/14-align-self/readme.md:
```diff
@@ -0,0 +1,13 @@
+With `align-items` you move all the items at once (on the **non-main axis**).
+With `align-self` you move only one of the items.
+
+![Example of align-items vs. align-self](./ignore_this_folder/example1.png)
```
I drew this up to try and illustrate them against the axes.

![flexbox-axis-align](https://user-images.githubusercontent.com/2306629/50686882-f4915180-1016-11e9-82d2-b0a374294676.png)

I've kept the source file for it, so if you want me to draw up any more for describing the main axis/cross axis, just sketch out something quick and I can reproduce it. :+1:

rarmatei replied:
Much better than my example 😃

NateWr commented on week-2/14-align-self/readme.md:
```diff
@@ -0,0 +1,13 @@
+
+You can use the same values for `align-self` as you used for `align-items`.
+
+Exercise:
```
Again, I'd say move this above the example image, just after the `align-items`, `align-self` descriptions. Then some more text like this:

> In this exercise, you will use `align-self` and `align-items` to change the position of the items. Follow the steps below to complete this exercise.
>
> 1. .... [like the description above]

NateWr commented on week-2/13-order/index.html:
```diff
@@ -0,0 +1,56 @@
+
+<head>
+  <meta charset="utf-8">
+  <title>Flexbox</title>
```
Also, I've been using the exercise number in the title, because I think students often have so many tabs open and struggle to find what they're working on. Something like:

> 13 - Flex Order - HTML, CSS and Git Exercises

And same for 14 below.

rarmatei commented:
Thanks @NateWr for the very thorough review - I fixed/responded to all of your comments.

Regarding the formatting issues - once all the currently open PRs are merged in, I'll add one to format everything using prettier and add it a "recommended" extension so everyone gets prompted to install it when they open the project in VS Code.

Also, I moved the solution images outside of each exercise and into a separate panel. I tried to keep to the side if the laptop is 13' or larger, otherwise it goes below. If you're ok with how it looks and you merge this PR, I can then do some later improvements and try to align each solution image to its exercise on the left.

NateWr commented:
> Also, I moved the solution images outside of each exercise and into a separate panel. I tried to keep to the side if the laptop is 13' or larger, otherwise it goes below. If you're ok with how it looks and you merge this PR, I can then do some later improvements and try to align each solution image to its exercise on the left.

This looks good. Let's leave it as-is for the first review. I'll work on the visual distinction between left and right in time for the class to start. But we've got the exercise, and the approach, in time for a third party to go through it for clarity/accuracy. :+1:
